File: windows/Services/HotKeyManager.cs
using WSNH.Models;

namespace WSNH.Services;

/// <summary>Registers/unregisters global hotkeys for each saved PromptAction.</summary>
public sealed class HotKeyManager
{
    public static readonly HotKeyManager Shared = new();

    private readonly Dictionary<Guid, int> _registeredIds = new();
    public Action<PromptAction>? OnTrigger { get; set; }

    private HotKeyManager() { }

    public void RegisterAll(IEnumerable<PromptAction> actions)
    {
        foreach (var id in _registeredIds.Values)
        {
            HotKeyRegistry.Shared.Unregister(id);
        }
        _registeredIds.Clear();

        foreach (var action in actions)
        {
            Register(action);
        }
    }

    public void Register(PromptAction action)
    {
        if (!action.HasHotkey) return;

        var registeredId = HotKeyRegistry.Shared.Register(action.KeyCode, action.ModifierFlags, () =>
        {
            OnTrigger?.Invoke(action);
        });

        if (registeredId is int id)
        {
            _registeredIds[action.Id] = id;
        }
    }

    public void Unregister(Guid actionId)
    {
        if (_registeredIds.TryGetValue(actionId, out var id))
        {
            HotKeyRegistry.Shared.Unregister(id);
            _registeredIds.Remove(actionId);
        }
    }
}

File: mac/Sources/WSNH/Services/HotKeyManager.swift
import AppKit
import HotKey

/// Registers/unregisters global hotkeys for each saved PromptAction using the
/// HotKey package (a thin wrapper around Carbon's RegisterEventHotKey).
final class HotKeyManager {
    static let shared = HotKeyManager()

    private var hotKeys: [UUID: HotKey] = [:]
    var onTrigger: ((PromptAction) -> Void)?

    private init() {}

    func registerAll(_ actions: [PromptAction]) {
        hotKeys.removeAll()
        for action in actions {
            register(action)
        }
    }

    func register(_ action: PromptAction) {
        guard let key = Key(carbonKeyCode: action.keyCode) else { return }
        let modifiers = NSEvent.ModifierFlags(rawValue: action.modifierFlags)
        let hotKey = HotKey(key: key, modifiers: modifiers)
        hotKey.keyDownHandler = { [weak self] in
            self?.onTrigger?(action)
        }
        hotKeys[action.id] = hotKey
    }

    func unregister(_ actionId: UUID) {
        hotKeys.removeValue(forKey: actionId)
    }
}
